tutel/experts/ffn.py

```diff
@@ -18,34 +23,35 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
             activation_fn = lambda x: F.relu(x)
         self.activation_fn = activation_fn
 
-    def update(self, ctx):
-        if ctx.sharded_count > 1:
-            assert self.hidden_size_per_expert % ctx.sharded_count == 0, f"Can't evenly divide hidden_size_per_expert ({self.hidden_size_per_expert}) to {ctx.sharded_count} slices."
-
-        hidden_size = self.hidden_size_per_expert // ctx.sharded_count
-        model_dim = ctx.model_dim
-        local_experts = ctx.num_local_experts
-        self.output_dim = self.output_dim or model_dim
-
-        fc1_weight = torch.empty(1, local_experts, hidden_size, model_dim)
-        fc2_weight = torch.empty(1, local_experts, hidden_size, self.output_dim)
-        fc1_bias = torch.empty(1, local_experts, hidden_size)
-        fc2_bias = torch.empty(1, local_experts, (self.output_dim + ctx.sharded_count - 1) // ctx.sharded_count)
-
-        for i in range(local_experts):
-            fc1 = torch.nn.Linear(model_dim, hidden_size)
-            fc2 = torch.nn.Linear(hidden_size, self.output_dim)
-            fc1_weight[0, i, :, :], fc1_bias[0, i, :] = fc1.weight, fc1.bias
-            fc2_weight[0, i, :, :], fc2_bias[0, i, :] = fc2.weight.t(), fc2.bias[:fc2_bias.size(-1)]
-
-        self.register_parameter(name='batched_fc1_w', param=torch.nn.Parameter(fc1_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc2_w', param=torch.nn.Parameter(fc2_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc1_bias', param=torch.nn.Parameter(fc1_bias.squeeze(0)))
-        self.register_parameter(name='batched_fc2_bias', param=torch.nn.Parameter(fc2_bias.squeeze(0)))
+        self.batched_fc1_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, model_dim))
+        self.batched_fc2_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, self.output_dim))
+        if has_fc1_bias:
+             self.batched_fc1_bias = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size))
```

Comment:
What is the reason not using `self.register_parameter('batched_fc1_bias', param=torch.nn.Parameter(torch.empty(local_experts, self.hidden_size)))`?

Reply:
param instantiation follows PyTorch convention eg [nn.Linear](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/linear.py#L96)

Let me know if you want to break from PyTorch convention

Reply:
OK, seems like there is tiny difference when paramter definitions are outside __init__ function (https://discuss.pytorch.org/t/nn-module-add-new-parameter-setattr-vs-register-parameter/108238). But inside __init__ function, they seem to be the same result set.